### PR TITLE
Fix STATICCALL

### PR DESCRIPTION
**Pull request by chfast**
Fixes https://github.com/ethereum/cpp-ethereum/issues/4199.

### CODE REVIEW

**Comment by chfast** [comment]
This also requires JSON tests update. I can't do it because I have issues with compiling LLL code.

File: libevm/VMCalls.cpp [hunk 1]
```diff
@@ -211,15 +211,17 @@ bool VM::caseCallSetup(CallParameters *callParams, bytesRef& o_output)
...
 	if (m_OP == Instruction::CALL && !m_ext->exists(destinationAddr))
 		if (m_SP[2] > 0 || m_schedule->zeroValueTransferChargesNewAccountGas())
 			m_runGas += toInt63(m_schedule->callNewAccountGas);
 
-	if ((m_OP == Instruction::CALL || m_OP == Instruction::CALLCODE) && m_SP[2] > 0)
+	if (haveValueArg && m_SP[2] > 0)
 		m_runGas += toInt63(m_schedule->callValueTransferGas);
 
-	size_t sizesOffset = (m_OP == Instruction::DELEGATECALL || m_OP == Instruction::STATICCALL) ? 2 : 3;
+	size_t const sizesOffset = haveValueArg ? 3 : 2;
```
Comment (chriseth) [inline libevm/VMCalls.cpp]: Did you forget a `!` here or was there a bug before?

File: libevm/VMCalls.cpp [hunk 2]
```diff
@@ -248,27 +250,26 @@ bool VM::caseCallSetup(CallParameters *callParams, bytesRef& o_output)
...
-		callParams->apparentValue = m_ext->value;
-		callParams->valueTransfer = 0;
+		callParams->valueTransfer = m_SP[2];
+		callParams->apparentValue = m_SP[2];
 	}
 	else
 	{
-		callParams->apparentValue = callParams->valueTransfer = m_SP[2];
-		inOutOffset = 1;
+		callParams->apparentValue = m_ext->value;
```
Comment (chriseth) [inline libevm/VMCalls.cpp]: Do we want to forward the value in case of staticcall? I would say thiscase should only apply to delegatecall.

File: libevm/VMCalls.cpp (hunk 1 above)
Comment (chfast) [inline libevm/VMCalls.cpp]: I also swapped the 2 and 3.

File: libevm/VMCalls.cpp (hunk 2 above)
Comment (chfast) [inline libevm/VMCalls.cpp]: You are probably right. In case of STATICALL the value should not be forwarded and be 0 all the time.
This might be another bug. @winsvega Can you look for some test cases for this?

Comment (winsvega) [inline libevm/VMCalls.cpp]: I have those on the list
actually callvalue has some strange value at this test 
https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stStaticCall/static_callToDelCallOpCodeCheckFiller.json#L43

Comment (winsvega) [inline libevm/VMCalls.cpp]: and here 
https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stStaticCall/static_callToStaticOpCodeCheckFiller.json#L43

Comment (chfast) [inline libevm/VMCalls.cpp]: All of them are quite complex. We need a simple test with a transaction having no-zero value, then staticcall and check of the callvalue inside this staticcall.

Similar one with a CALL instead of transaction.

Comment (winsvega) [inline libevm/VMCalls.cpp]: this test does exactly just that
just add more values to the value list

Comment (chfast) [inline libevm/VMCalls.cpp]: I'm not sure what you mean by "this test". You referenced 2 tests before. If you mean https://github.com/ethereum/tests/blob/develop/src/GeneralStateTestsFiller/stStaticCall/static_callToStaticOpCodeCheckFiller.json#L43, this tests does other weird things like: checking for ORIGIN, multiple levels of different calls, etc. I'm not sure any of them check if the value from transaction is forwarded in STATICCALL.

Also, I don't know what you mean by "just add more values to the value list".

Comment (pirapira) [inline libevm/VMCalls.cpp]: I will add tests.

1. call transaction into A with non-zero value; A performs STATICCALL to B; B stores the call value
2. call transaction into Q; Q calls into A with non-zero value; A performs STATICCALL to B; B stores the call value

In both cases, the storage slot of `B` would initially contain some non-zero value, so that we can be sure that the value is overwritten by zero.

**Comment by pirapira** [comment]
I added the two tests.

**Comment by pirapira** [comment]
Let me have a look at these
* `static_callToDelCallOpCodeCheck Expected another postState hash`
* `static_callToStaticOpCodeCheck Expected another postState hash`

**Comment by chfast** [comment]
Ready to go?

**Comment by pirapira** [comment]
@winsvega sounds not happy about the tests I changed.

**Comment by pirapira** [comment]
I changed the tests and rendered them useless.  I will apply correct fixes.
